UART Driver/main.c

```diff
@@ -0,0 +1,307 @@
+  * in the root directory of this software component.
+  * If no LICENSE file comes with this software, it is provided AS-IS.
+  *
+  ******************************************************************************
+  */
+/* USER CODE END Header */
+/* Includes ------------------------------------------------------------------*/
+#include "main.h"
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+UART_HandleTypeDef huart2;
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+static void MX_GPIO_Init(void);
+static void MX_USART2_UART_Init(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+//Creates the fijo and foji struct
+//The reason there are 2 different gpsCoordinates structs is to minimize the size of the fijo struct
+struct gpsCoordinatesFOJI{
+	double lattitude;
+	double longtitude;
+	double altitude;
+};
+
+struct gpsCoordinatesFIJO{
+	double lattitude;
+	double longtitude;
+};
+
+struct eulerAnglesOfPlane{
+	float yaw;
+	float pitch;
+	float roll;
+};
+
+//Should be 37 bytes but because of padding its really 48 bytes
+ struct foji{
+	uint8_t start;
+	struct gpsCoordinatesFOJI gpsCoord;
+	struct eulerAnglesOfPlane anglesOfPlane;
+};
+
+//Should be 20 bytes but because of padding its really 24 bytes
+struct fijo{
+	 uint8_t start;
+	_Bool takeoffCommand;
+	_Bool qrScanFlag;
+	_Bool detectFlag;
+	struct gpsCoordinatesFIJO gpsCoord;
+};
+
+const uint8_t STARTBYTE = 0x24;
+
+void sendFOJI(){
+	//Get out from telemetry manager
+	struct foji out;
+
+	HAL_UART_Transmit(&huart2, &out, sizeof(struct foji), HAL_MAX_DELAY);
+
+}
+
+void readFIJO(){
+	//Creates byte array
+	uint8_t buf[sizeof(struct fijo)];
+
+	//Initializes starting byte
+	buf[0] = STARTBYTE;
+
+	struct fijo in;
+	in.start = STARTBYTE;
+
+	//Receives a byte to check if its the same as STARTBYTE
+	uint8_t check;
+	HAL_UART_Receive(&huart2, &check, 1, HAL_MAX_DELAY);
+
+	if(check == STARTBYTE){
+		//Receive all the message bytes in buf
+		for(uint16_t i = 1; i < sizeof(struct fijo); i++)
+		{
+			HAL_UART_Receive(&huart2, &buf[i], 1, HAL_MAX_DELAY);
+		}
```

Comment:
I believe that there is a way to read in a number of bytes without needing a loop. Think you pass in the sizeof(struct fijo) instead of a 11 in the UART_Receive function, but that would be more ideal than using a loop. Be careful about whether the data coming in is stored in little endian or big endian format (order of bytes).

Reply:
I was thinking about this too but in this case, I would need to start receiving from index 1 and not 0. So if I put the address of my array and the size of the struct as my parameter it will mess it up cuz it starts from the 0 index

Reply:
Could you add a single byte to your struct of padding? It seems like that byte gets transmitted anyway.

Reply:
Hmm, I am not sure I understand what you mean. But do you mean like adding it when I cast my byte array to a struct?

Reply:
Adding it at the start to solve that indexing issue. Then the data you care about can be read from the struct and that one byte will be ignored.

Reply:
oh like getting rid of the "start" uint8_t from the fijo struct altogether because it's not functional? that would work